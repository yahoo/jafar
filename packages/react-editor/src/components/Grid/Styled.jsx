/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import styled from 'styled-components';
import Button from '@material-ui/core/Button';

export const HeaderButton = styled(Button)`
  margin-left: 10px !important;
`;

export const Grid = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  overflow: hidden;
`;

export const Row = styled.div`
  display: flex;
  flex-direction: row;
  line-height: 40px;
  [aria-label="actions"] {
    visibility: hidden;
  }
  &:hover {
    [aria-label="actions"] {
      visibility: visible;
    }
  }
  &:nth-child(even) {
    background-color: #f9f9f9;
  }
  &:nth-child(odd) {
    background-color: #f1f1f1;
  }
`;

export const HeaderRow = styled(Row)`
  font-weight: bold;
  background-color: #e0e0e0 !important;
`;

export const Cell = styled.div`
  padding: 10px 20px;
  flex: ${props => props.minWidth ? 'unset' : 1 };
  min-width: ${props => props.minWidth};
  overflow: hidden;
`;

export const Body = styled.div`
  flex: 1;
  display: flex;
  flex-direction: column;
  overflow-y: auto;
`;

export const HeaderMenu = styled.div`
  text-align: right;
  padding-bottom: 10px;
`;
