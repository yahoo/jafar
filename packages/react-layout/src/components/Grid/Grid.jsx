/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import styled from 'styled-components';

const Grid = styled.div`
  display: grid;
  grid-template-areas: ${props => props.templateAreas.map(row => `"${row}"`).join(' ')};

  ${props => props.elements
    .map(element => `${element.selector} { grid-area: ${element.gridArea}; ${element.style || ''} }`)
    .join(' ')}
`;

export default ({ templateAreas, elements }) => (<Grid templateAreas={templateAreas} elements={elements}>
  {
    elements.map((element, index) => {
      const GenericComponent = element.component;
      return <GenericComponent key={index} {...element.props} />;
    })
  }
</Grid>);
