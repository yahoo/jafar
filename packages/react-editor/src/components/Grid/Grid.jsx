/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import ActionsMenu from '../ActionsMenu';
import * as Styled from './Styled';

const Grid = ({ columns, data, headerActions, rowActions }) => {  
  return (<Styled.Grid aria-label="grid">
    <Styled.HeaderMenu aria-label="grid-header-menu">
      {
        headerActions.map((action, index) => (<Styled.HeaderButton key={index} variant={action.variant || 'contained'} 
          color="primary" onClick={action.onClick}>{action.label}</Styled.HeaderButton>))
      }
    </Styled.HeaderMenu>
    <Styled.HeaderRow>
      {
        columns.map((column, index) => (<Styled.Cell key={index}>{column.label}</Styled.Cell>))
      }
      {
        rowActions && <Styled.Cell minWidth="50px" />
      }
    </Styled.HeaderRow>
    <Styled.Body aria-label="grid-body">
      {
        data.map((row, index) => (<Styled.Row key={index} aria-label="grid-row">
          {
            columns.map((column, index2) => (<Styled.Cell key={index2}>{column.content(row)}</Styled.Cell>))
          }
          {
            rowActions && <Styled.Cell minWidth="50px">
              <ActionsMenu data={row} options={rowActions} />
            </Styled.Cell>
          }
        </Styled.Row>))
      }
    </Styled.Body>
  </Styled.Grid>);
};

export default Grid;
