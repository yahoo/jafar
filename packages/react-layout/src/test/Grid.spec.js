/**
  * Copyright 2020, Verizon Media
  * Licensed under the terms of the MIT license. See LICENSE file in project root for terms.
  */

import React from 'react';
import { shallow } from 'enzyme';
import Grid from '../components/Grid/index';

describe('Grid', () => {
  const Field = () => {};

  it('Should render grid', () => {
    const props = getGrid([
      'firstName lastName .',
      'personalId address .',
    ]);

    const component = shallow(<Grid {...props} />);
    expect(component).toMatchSnapshot();
  });

  function getGrid(templateAreas) {
    const fieldIds = templateAreas.join(' ').split(' ').filter(x => x !== '.');
    return {
      templateAreas,
      elements: fieldIds.map(id => ({ 
        selector: `#${id}`, 
        gridArea: id, 
        component: Field, 
        props: { id },
        style: 'width: 350px;',
      })),
    };
  }
});
