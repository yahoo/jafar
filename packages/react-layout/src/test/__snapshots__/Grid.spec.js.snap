// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Grid Should render grid 1`] = `
<styled.div
  elements={
    Array [
      Object {
        "component": [Function],
        "gridArea": "firstName",
        "props": Object {
          "id": "firstName",
        },
        "selector": "#firstName",
        "style": "width: 350px;",
      },
      Object {
        "component": [Function],
        "gridArea": "lastName",
        "props": Object {
          "id": "lastName",
        },
        "selector": "#lastName",
        "style": "width: 350px;",
      },
      Object {
        "component": [Function],
        "gridArea": "personalId",
        "props": Object {
          "id": "personalId",
        },
        "selector": "#personalId",
        "style": "width: 350px;",
      },
      Object {
        "component": [Function],
        "gridArea": "address",
        "props": Object {
          "id": "address",
        },
        "selector": "#address",
        "style": "width: 350px;",
      },
    ]
  }
  templateAreas={
    Array [
      "firstName lastName .",
      "personalId address .",
    ]
  }
>
  <Field
    id="firstName"
    key="0"
  />
  <Field
    id="lastName"
    key="1"
  />
  <Field
    id="personalId"
    key="2"
  />
  <Field
    id="address"
    key="3"
  />
</styled.div>
`;
